### PR TITLE
HIVE-25397: snapshot support for controlled failover

### PR DESCRIPTION
**Pull request by ArkoSharma**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 1]
```diff
@@ -192,54 +192,130 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
-  private SnapshotUtils.SnapshotCopyMode createSnapshotsAtSource(Path sourcePath, String snapshotPrefix,
+  private SnapshotUtils.SnapshotCopyMode createSnapshotsAtSource(Path sourcePath, Path targetPath, String snapshotPrefix,
       boolean isSnapshotEnabled, HiveConf conf, SnapshotUtils.ReplSnapshotCount replSnapshotCount, FileList snapPathFileList,
       ArrayList<String> prevSnaps, boolean isBootstrap) throws IOException {
     if (!isSnapshotEnabled) {
       LOG.info("Snapshot copy not enabled for path {} Will use normal distCp for copying data.", sourcePath);
       return FALLBACK_COPY;
     }
     DistributedFileSystem sourceDfs = SnapshotUtils.getDFS(sourcePath, conf);
+    DistributedFileSystem targetDfs = SnapshotUtils.getDFS(targetPath, conf);
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:203]: we can not use the targetFS, during the dump, it won't be accessible in some cases. We should use targetFS only during copy phase

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 2]
```diff
@@ -192,64 +196,135 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+        } catch (SnapshotException e) {
+          //dir not snapshottable, continue
+        }
+      }
+      boolean isFirstSnapshotAvl =
+              SnapshotUtils.isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix, OLD_SNAPSHOT, conf);
+      boolean isSecondSnapAvl =
+              SnapshotUtils.isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix, NEW_SNAPSHOT, conf);
+      //for bootstrap and non - failback case, use initial_copy
+      if(isBootstrap && !(!isSecondSnapAvl && isFirstSnapshotAvl)) {
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:238]: Is this all to find out. the prefix, which is always the database name?  This seems to be done for even incremental also now?
Can we just not rename the snapshot in case of control failover as per the new standards and use the normal flow post that and get rid of maintaining prefix

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 3]
```diff
@@ -103,38 +109,49 @@ static void internalBeforeClassSetup(Map<String, String> overrides,
...
+      put("hive.txn.manager", "org.apache.hadoop.hive.ql.lockmgr.DbTxnManager");
+      put("hive.metastore.client.capability.check", "false");
+      put("hive.repl.bootstrap.dump.open.txn.timeout", "1s");
+      put("hive.strict.checks.bucketing", "false");
+      put("hive.mapred.mode", "nonstrict");
+      put("mapred.input.dir.recursive", "true");
+      put("hive.metastore.disallow.incompatible.col.type.changes", "false");
+      put("metastore.warehouse.tenant.colocation", "true");
+      put("hive.in.repl.test", "true");
+      put("hive.txn.readonly.enabled", "true");
```
Comment (ayushtkn) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:136]: nit
looks like only indentation change, avoid unrelated changes

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosUsingSnapshots.java [hunk 4]
```diff
@@ -79,11 +80,11 @@ public static void classLevelSetup() throws Exception {
     overrides.put(HiveConf.ConfVars.REPL_INCLUDE_EXTERNAL_TABLES.varname, "true");
     overrides.put(HiveConf.ConfVars.HIVE_DISTCP_DOAS_USER.varname,
         UserGroupInformation.getCurrentUser().getUserName());
-    overrides.put(HiveConf.ConfVars.REPL_RUN_DATA_COPY_TASKS_ON_TARGET.varname, "false");
+    overrides.put(HiveConf.ConfVars.REPL_RUN_DATA_COPY_TASKS_ON_TARGET.varname, "true");
     overrides.put(HiveConf.ConfVars.REPL_EXTERNAL_WAREHOUSE_SINGLE_COPY_TASK.varname, "true");
     overrides.put(REPL_SNAPSHOT_DIFF_FOR_EXTERNAL_TABLE_COPY.varname, "true");
 
-    internalBeforeClassSetup(overrides, TestReplicationScenarios.class);
+    internalBeforeClassSetupExclusiveReplica(overrides, overrides, TestReplicationScenarios.class);
```
Comment (ayushtkn) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosUsingSnapshots.java:87]: why do we need this? all the test should work as is without bothering existing tests.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 5]
```diff
@@ -192,64 +196,135 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
         SnapshotUtils.deleteSnapshotIfExists(sourceDfs, sourcePath, firstSnapshot(snapshotPrefix), conf);
         SnapshotUtils.deleteSnapshotIfExists(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf);
         allowAndCreateInitialSnapshot(sourcePath, snapshotPrefix, conf, replSnapshotCount, snapPathFileList, sourceDfs);
-        return INITIAL_COPY;
+        ret.put(prefix, INITIAL_COPY);
+        return ret;
       }
 
+      //While resuming a failed replication
       if (prevSnaps.contains(sourcePath.toString())) {
```
Comment (ayushtkn) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:248]: is prevSnaps initialised for bootstrap? better to add a test as well for the failure and checkpointing flow

**Review by ayushtkn — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java [hunk 6]
```diff
@@ -243,18 +244,25 @@ boolean copyUsingDistCpSnapshots(Path sourcePath, Path targetPath, UserGroupInfo
 
     DistributedFileSystem targetFs = SnapshotUtils.getDFS(targetPath, clonedConf);
     boolean result = false;
+    boolean isSecondSnapshotAvl =
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:247]: rename to more readable variables

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java [hunk 7]
```diff
@@ -243,18 +244,25 @@ boolean copyUsingDistCpSnapshots(Path sourcePath, Path targetPath, UserGroupInfo
...
+            SnapshotUtils.isSnapshotAvailable(targetFs, targetPath, work.getSnapshotPrefix(), NEW_SNAPSHOT, clonedConf);
     if (getWork().getCopyMode().equals(SnapshotUtils.SnapshotCopyMode.DIFF_COPY)) {
+      if(isSecondSnapshotAvl) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:250]: add comments why this is done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java [hunk 8]
```diff
@@ -270,8 +278,9 @@ boolean copyUsingDistCpSnapshots(Path sourcePath, Path targetPath, UserGroupInfo
       // snapshots.
       SnapshotUtils.allowSnapshot(targetFs, work.getFullyQualifiedTargetPath(), clonedConf);
       // Attempt to delete the snapshot, in case this is a bootstrap post a failed incremental, Since in case of
-      // bootstrap we go from start, so delete any pre-existing snapshot.
+      // bootstrap we go from start, so delete any pre-existing snapshot, (both snapshots can exist in case of failback)
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:281]: do you mean failover?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java (hunk 2 above)
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:238]: Considered making this change, but realised this would then require a similar listing and name-checking for snapshot on the target side. This can occur in case of reverse replication after failover with different names of src and tgt dbs.
Hence decided to proceed with this implementation itself as it allows for identifying which snapshots are being re-used. Also makes sense to include this work with dump considering that in general, dump should take lesser time than load (except possibly cases with just external tables in db with data-copy on src).

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosUsingSnapshots.java (hunk 4 above)
Comment (ArkoSharma) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosUsingSnapshots.java:87]: The tests are modified to use different filesystems for src and tgt. A few wrongFS related errors were fixed and this required testing with different fs for src and tgt.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java (hunk 5 above)
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:248]: Added.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java (hunk 6 above)
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:247]: Renamed.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java (hunk 7 above)
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:250]: Added.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java (hunk 8 above)
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java:281]: Changed to read 'reverse replication'

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 9]
```diff
@@ -189,57 +191,137 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+          //this can be used in re-bootstrap cases after irrecoverable error
+          if(SnapshotUtils.deleteSnapshotIfExists(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf)) {
+            replSnapshotCount.incrementNumDeleted();
+          }
+          SnapshotUtils.createSnapshot(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf);
+          snapPathFileList.add(sourcePath.toString());
+          replSnapshotCount.incrementNumCreated();
+          return SnapshotUtils
+                  .isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix, OLD_SNAPSHOT, conf) ? DIFF_COPY : INITIAL_COPY;
+        } else {
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:260]: This case is included in the following block.

(!secondSnapAvailable && firstSnapAvailable) - this condition denotes reverse replication taking place for the first time for a particular path. So the execution is similar for both incremental and bootstrap - whereby we either reuse the snaps or don't depending upon the conf. For this purpose, this case is clubbed for both.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 10]
```diff
@@ -189,57 +191,137 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+      if(isBootstrap && conf.getBoolVar(HiveConf.ConfVars.REPL_REUSE_SNAPSHOTS)) {
+        try {
+          FileStatus[] listing = sourceDfs.listStatus(new Path(sourcePath, ".snapshot"));
+          for (FileStatus elem : listing) {
+            String snapShotName = elem.getPath().getName();
+            String prefix;
+            if (snapShotName.contains(OLD_SNAPSHOT)) {
+              prefix = snapShotName.substring(0, snapShotName.lastIndexOf(OLD_SNAPSHOT));
+              if(!prefix.equals(snapshotPrefix)) {
+                sourceDfs.renameSnapshot(sourcePath, firstSnapshot(prefix), firstSnapshot(snapshotPrefix));
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java]: done.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 11]
```diff
@@ -189,57 +191,137 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+              }
+            }
+            if (snapShotName.contains(NEW_SNAPSHOT)) {
+              prefix = snapShotName.substring(0, snapShotName.lastIndexOf(NEW_SNAPSHOT));
+              if(!prefix.equals(snapshotPrefix)) {
+                sourceDfs.renameSnapshot(sourcePath, secondSnapshot(prefix), secondSnapshot(snapshotPrefix));
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java]: done.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 12]
```diff
@@ -189,57 +191,137 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+        // In case of reverse replication after fail-over, in some paths, second snapshot may not be present.
+        if(SnapshotUtils.deleteSnapshotIfExists(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf)) {
+          replSnapshotCount.incrementNumDeleted();
+        }
         SnapshotUtils.createSnapshot(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf);
         replSnapshotCount.incrementNumCreated();
         snapPathFileList.add(sourcePath.toString());
         return SnapshotUtils
-            .isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix, OLD_SNAPSHOT, conf) ? DIFF_COPY : INITIAL_COPY;
+                .isSnapshotAvailable(sourceDfs, sourcePath, snapshotPrefix, OLD_SNAPSHOT, conf) ? DIFF_COPY : INITIAL_COPY;
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java]: done.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 13]
```diff
@@ -189,57 +191,137 @@ private void dirLocationToCopy(String tableName, FileList fileList, Path sourceP
...
+          sourceDfs.deleteSnapshot(sourcePath, firstSnapshot(snapshotPrefix));
+          replSnapshotCount.incrementNumDeleted();
+          SnapshotUtils.renameSnapshot(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), firstSnapshot(snapshotPrefix), conf);
+          SnapshotUtils.createSnapshot(sourceDfs, sourcePath, secondSnapshot(snapshotPrefix), conf);
+          replSnapshotCount.incrementNumCreated();
+          snapPathFileList.add(sourcePath.toString());
+          return DIFF_COPY;
+        } else {
+          //only new snapshot exists
           sourceDfs.renameSnapshot(sourcePath, secondSnapshot(snapshotPrefix), firstSnapshot(snapshotPrefix));
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java]: done.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 14]
```diff
@@ -1105,17 +1105,15 @@ Long bootStrapDump(Path dumpRoot, DumpMetaData dmd, Path cmRoot, Hive hiveDb)
...
 
           String snapshotPrefix = dbName.toLowerCase();
-          ArrayList<String> prevSnaps = new ArrayList<>(); // Will stay empty in case of bootstrap
+          ArrayList<String> prevSnaps = new ArrayList<>();
           if (isSnapshotEnabled) {
-            // Delete any old existing snapshot file, We always start fresh in case of bootstrap.
             FileUtils.deleteIfExists(getDFS(SnapshotUtils.getSnapshotFileListPath(dumpRoot), conf),
-                new Path(SnapshotUtils.getSnapshotFileListPath(dumpRoot),
-                    EximUtil.FILE_LIST_EXTERNAL_SNAPSHOT_CURRENT));
-            FileUtils.deleteIfExists(getDFS(SnapshotUtils.getSnapshotFileListPath(dumpRoot), conf),
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:1114]: made the change.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java [hunk 15]
```diff
@@ -254,23 +256,55 @@ public boolean canExecuteInParallel() {
...
+            targetFs.renameSnapshot(targetPath, firstSnapshot(prefix), firstSnapshot(snapPrefix));
+          }
+        }
+        if (snapShotName.contains(NEW_SNAPSHOT)) {
+          prefix = snapShotName.substring(0, snapShotName.lastIndexOf(NEW_SNAPSHOT));
+          if (!prefix.equals(snapPrefix)) {
+            targetFs.renameSnapshot(targetPath, secondSnapshot(prefix), secondSnapshot(snapPrefix));
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java]: done.

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java [hunk 16]
```diff
@@ -1237,25 +1238,42 @@ public boolean runDistCpWithSnapshots(String oldSnapshot, String newSnapshot, Li
...
+            // Get the path relative to the initial snapshot for copy.
+            Path snapRelPath = new Path(srcPaths.get(0), HdfsConstants.DOT_SNAPSHOT_DIR + "/" + newSnapshot);
+            // Copy from the initial snapshot path.
+            if(runDistCp(Collections.singletonList(snapRelPath), dst, conf)) {
               return true;
             } else {
-              LOG.error("Copy failed with after target restore for source: {} to target: {} snapshot1: {} snapshot2: "
-                  + "{} params: {}. Return code: {}", srcPaths, dst, oldSnapshot, newSnapshot, params, returnCode);
+              LOG.error("Copy failed with full distcp for source: {} to target: {}  snapshot:"
+                      + "{} params: {}. Return code: {}", srcPaths, dst, newSnapshot, returnCode);
```
Comment (ayushtkn) [inline shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:1276]: Incorrect number of placeholders. Misses params

**Review by ayushtkn — APPROVED** [review]
Minor nits.
Apart Changes LGTM

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 17]
```diff
@@ -695,6 +695,10 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
         + "data copy, the target data is overwritten and the modifications are removed and the copy is again "
         + "attempted using the snapshot based approach. If disabled, the replication will fail in case the target is "
         + "modified."),
+    REPL_REUSE_SNAPSHOTS("hive.repl.reuse.snapshots", false,
+        "If enabled,reusing snapshots is attempted in case of controlled failover(B->A) when same paths are"
+        + "used for external table replication on src and target. Also in cases of failed incremental where re-bootstrap is required."
+        + "If set to true and snapshots exist in some paths, it creates/reuses new snapshots in those paths using the same name as exisiting snapshots."),
```
[the remark this answers is not in the document]
Comment (ArkoSharma) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java]: Corrected

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java (hunk 16 above)
Comment (ArkoSharma) [reply shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java]: Corrected
